### DISCUSSION TITLE
Peering based on libp2p

### DISCUSSION THREAD

**Discussion by iacore** [Kindelia/Kindelia#213] [Ideas]
libp2p is a mature p2p library, handling peering and communication over UDP DHT, WebSocket, and some other other transports. It has a Rust implementation.

I think it's a good idea for Kindelia to use this library to find peers and to communicate with peers.

If using libp2p and its own network isn't enough, there's also IPFS pubsub, which uses IPFS clients as message relays.

**Comment by steinerkelvin** [comment]
I believe this should definitely be done, just in a different layer/module then that which we are propagating tip blocks, which is pure UDP so it's fast as possible. Also, it keeps the base protocol as simple as possible, which is one of our goals.

Peer discovery and any other forms of coordination should definitely be done with something like `libp2p`.

Could you elaborate on IPFS pubsub?

We are probably going to use IPFS with the Kind-Lang-Kindelia toolchain to store source code (including proofs) that were compiled to Kindelia, it would be nice if we can integrate all of this somehow.

**Comment by iacore** [comment]
I don't know about about the implementation details.

IPFS is built using libp2p.
IPFS pubsub is built on libp2p, with its own peering parameters. IPFS pubsub also uses existing IPFS network for message passing.

The current Go libp2p pubsub is gossip based. I don't know if this is fast enough to reach consistency for a block-chain (used by Kindelia).

IPFS pubsub (Go implementation): https://blog.ipfs.tech/2020-05-20-gossipsub-v1.1/
libp2p docs: https://docs.libp2p.io/concepts/publish-subscribe/
